We will see here how to connect to an API without the help of a wrapper package such as `tweepy`. We will use the example of Twitter, but this should give you an idea about requesting data from APIs more generally - it is essentially like requesting a website! Hence, we will use `requests`. If you want to learn more, I recommend this blogpost to get started: https://www.dataquest.io/blog/python-api-tutorial/

In [1]:
import requests
import json
import pandas as pd

On the Twitter Developer Platform under https://developer.twitter.com/en/docs/api-reference-index#Twitter, you can find the different API "endpoints" that Twitter provides (essentially, there is a different API depending on what data you want). We will be using here version 1.1 and searching for tweets. Hence, we follow the link https://developer.twitter.com/en/docs/twitter-api/v1/tweets/search/api-reference/get-search-tweets. Here, we find a "Resource URL", which provides the relevant data: 'https://api.twitter.com/1.1/search/tweets.json' (essentially, the location of the API application on Twitter's server). We can send a request to this site, but we won't get much of a response:

In [2]:
requests.get("https://api.twitter.com/2/tweets/search/recent")

<Response [401]>

Why? because we haven't authenticated ourselves (response 401 indicates that we are not authorized). We can use the Consumer Key and Secret together with the Access Key and Secret, but it is easier to just use the Bearer Token:

In [3]:
BEARER_TOKEN = '[Add your Bearer token here]'

We use the token to form a "header", which tells the server who we are. https://developer.twitter.com/en/docs/twitter-api/v1/tweets/search/api-reference/get-search-tweets provides an example request (from Terminal, instead of Python), which gives us a rough idea of the required data:

In [4]:
headers = {
    "Authorization": "Bearer {}".format(BEARER_TOKEN),
}

We also specify parameters: this is what we are looking for? This corresponds to the query parameter we used in the `tweepy.Cursor`. At https://developer.twitter.com/en/docs/twitter-api/v1/tweets/search/api-reference/get-search-tweets, you can find all the possible parameters

In [5]:
parameters={
    'q': '#redbull',
}

We can now retry our request with the header and the search parameters:

In [6]:
response = requests.get("https://api.twitter.com/1.1/search/tweets.json",
                        headers = headers,
                        params = parameters)

In [7]:
response

<Response [401]>

This time, we get a better outcome: 200 indicates that the request was accepted by the server and we get a "normal" response. Of course, we can now print this out. We use the fact that the API returns information in JSON format (this is the case for most modern APIs. Older APIs tend to return XMLs).

In [8]:
response_data = response.json()
response_data

{'errors': [{'code': 89, 'message': 'Invalid or expired token.'}]}

This looks familiar, right? In fact, we can look at the typical tweet attributes we already learned about (just that we now have dictionary notation, instead of attributes)

In [9]:
response_data['statuses'][0]['text']

KeyError: 'statuses'

In [ ]:
response_data['statuses'][0]['user']['screen_name']